### PR TITLE
feat: add direct plugin framework

### PR DESCRIPTION
**Pull request by jsnwesson**
[APER-3108](https://2u-internal.atlassian.net/browse/APER-3108)
This PR formally adds Braden McDonald's plugin framework as a non-iFrame alternative for the library. The logic is still separated from the original framework's handling, so a follow-up ticket will be to integrate the two plugins -- iFrame and Direct. 

To see this PR running in Learner Dashboard: 
1. Pull branches down from Learner Dashboard
2. Checkout `aperture/add-UI-Plugin-Framework`
3. `npm install` and `npm run start`
4. The files with the Direct Plugin logic will be found in `/src/uiPlugins/SideBarSlot/index.jsx`, `src/uiPlugins/DemoPlugin/index.jsx`, and `src/containers/Dashboard/index.jsx`
[Changes can also be found here in GitHub](https://github.com/openedx/frontend-app-learner-dashboard/tree/aperture/add-UI-Plugin-Framework)

### CODE REVIEW

File: src/plugins/directPlugins/DirectPlugin.jsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+export const DirectPluginOperations = {
+  Insert: 'insert',
+  Hide: 'hide',
+  Modify: 'modify',
+  Wrap: 'wrap',
+};
+
+/**
+  This is what the allSlotChanges configuration should look like when passed into DirectPluginContext
+  {
```
Comment (jsnwesson) [inline src/plugins/directPlugins/DirectPlugin.jsx:28]: The one thing that I decided to change in order to steer away from Braden's approach and move closer to eventually using a JS-based config is that his configuration assumed that there'd be multiple `allSlotChanges` for a Host MFE, and I think whatever benefit that had could've eventually led to confusion/oversight.
I'd argue that keeping all known changes in this one object is easier to track, especially as they'll ideally live in a single JS config anyways.

File: src/plugins/directPlugins/mocks/DefaultComponentMock.jsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+const enabledPlugins = [
+  navLinksPlugin,
+];
+
+const MyApp = () => (
+  <DirectPluginContext value={enabledPlugins}>
+    <div>
+      <DirectPluginSlot
+        slotId="side-bar-nav"
+        defaultContents={navLinksPlugin.defaultLinks}
```
Comment (MaxFrank13) [inline src/plugins/directPlugins/mocks/DefaultComponentMock.jsx:19]: I have two questions regarding this `DefaultComponentMock` file. 

1. It doesn't appear to be used, yet. Is that correct?
2. The value `defaultLinks` doesn't exist as a key in the `navLinksPlugin` object. Is it supposed to be `defaultComponentProps`?

File: src/plugins/directPlugins/utils.test.jsx [hunk 3]
```diff
@@ -0,0 +1,175 @@
+import '@testing-library/jest-dom';
+
+import organizePlugins from './utils';
+import { DirectPluginOperations } from './DirectPlugin';
+
+jest.unmock('./utils');
```
Comment (MaxFrank13) [inline src/plugins/directPlugins/utils.test.jsx:6]: Curious because I've never used this before. What does the `jest.unmock` do for us here? Is `utils` being mocked somewhere else and this lets us use the actual file?

File: src/plugins/directPlugins/DirectPluginSlot.jsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  const allPluginChanges = React.useContext(DirectPluginContext);
+
+  const contents = React.useMemo(() => {
+    const slotChanges = allPluginChanges.getDirectSlotChanges()[slotId] ?? [];
+    return organizePlugins(defaultContents, slotChanges);
+  }, [allPluginChanges, defaultContents, slotId]);
+
+  return (
+    <>
+      {contents.map((c) => {
```
Comment (MaxFrank13) [inline src/plugins/directPlugins/DirectPluginSlot.jsx:17]: Something that was kind've throwing me off when we went over this was the use of `c` and `contents` as variable names. Maybe we can reword this to make it a little more clear what is actually happening here? I'm not sure exactly what the wording would be but would be down to meet up and workshop some ideas

File: src/plugins/directPlugins/utils.test.jsx [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  };
+  const modifiedWidget = widget;
+  modifiedWidget.content = newContent;
+  return modifiedWidget;
+};
+
+function mockWrapComponent({ widget }) {
+  const isAdmin = true;
+  return isAdmin ? widget : null;
+}
```
Comment (MaxFrank13) [inline src/plugins/directPlugins/utils.test.jsx:21]: Would it make more sense to call these `mockWrapOperation` and `mockModifyOperation` to better explain what they're doing?

File: src/plugins/directPlugins/mocks/PluginComponentsMock.jsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
+/* eslint-disable react/prop-types */
+// eslint-disable-next-line import/no-extraneous-dependencies
+import React from 'react';
+import {
+  House, Star, InsertDriveFile, Login,
+} from '@edx/paragon/icons';
+import { DirectPluginOperations } from '..';
+
+/** This is for us to be able to mock in tests */
+const isAdminHelper = () => true;
```
Comment (MaxFrank13) [inline src/plugins/directPlugins/mocks/PluginComponentsMock.jsx:10]: Can you better explain why this needed? I saw you were able to mock this behavior another way in the `utils.test` file

**Review by MaxFrank13 — COMMENTED** [review]
Thanks for putting this together. It looks great so far. I just had some questions to clear up things that were still hazy to me and a couple recommendations that were pretty minor.

File: src/plugins/directPlugins/utils.test.jsx (hunk 5 above)
Comment (jsnwesson) [reply src/plugins/directPlugins/utils.test.jsx:21]: great point on at least striving for clearer naming based on verbiage we already use, so I'm going to go with `mockModifyWidget` and `mockWrapWidget` since `mock...Operation` would imply that the operation itself is being mocked rather than the widget.

File: src/plugins/directPlugins/DirectPluginSlot.jsx (hunk 4 above)
Comment (jsnwesson) [reply src/plugins/directPlugins/DirectPluginSlot.jsx:17]: agreed on re-wording to add clarity rather than simplicity! I'd say `contents` is more accurately `preparedWidgets` and `c` is `preppedWidget`. Open to continue rethinking about it as we begin integrating this code with the iframe plugin

File: src/plugins/directPlugins/mocks/DefaultComponentMock.jsx (hunk 2 above)
Comment (jsnwesson) [reply src/plugins/directPlugins/mocks/DefaultComponentMock.jsx:19]: Originally I wrote these mock files before I truly started writing tests due to knowing I'd need to mock the config/implementations, but then realized that it was easier to start with mocking in the test file first. I think these files overall aren't necessary, and given we'll have plenty of docs/example apps to go off of, I'll remove these.

File: src/plugins/directPlugins/mocks/PluginComponentsMock.jsx (hunk 6 above)
Comment (jsnwesson) [reply src/plugins/directPlugins/mocks/PluginComponentsMock.jsx:10]: As mentioned before, will be removing the mock files due to their lack of need.

File: src/plugins/directPlugins/utils.test.jsx (hunk 3 above)
Comment (jsnwesson) [reply src/plugins/directPlugins/utils.test.jsx:6]: it actually doesn't do anything for us in this moment and I completely forgot to remove it. To your prediction, it would have been to ensure that if (and when) `/utils` is mocked in other test files via `setupTests.js`, then it would not automatically be mocked in this file. https://m-t-a.medium.com/jest-clearing-module-mocks-effortlessly-6656d104cab

**Review by MaxFrank13 — APPROVED** [review]
